pymess/backend/__init__.py

```diff
@@ -1,27 +1,130 @@
+
+    @transaction.atomic
+    def send(self, recipient, content, related_objects=None, tag=None, template=None, **kwargs):
+        """
+        Send message with the text content to the phone number (recipient)
+        :param recipient: email or phone number of the recipient
+        :param content: text content of the message
+        :param related_objects: list of related objects that will be linked with the message using generic
+        relation
+        :param tag: string mark that will be saved with the message
+        :param template: template object from which content of the message was create
+        :param kwargs: extra attributes that will be stored to the message
+        """
+        backend = self.get_backend(recipient)
+        message = self.create_message(recipient, content, related_objects, tag, template, **kwargs)
+        if not backend.is_turned_on_batch_sending():
+            backend.publish_message(message)
+        return message
+
+    def get_batch_size(self):
+        """
+        Return number of messages sent in batch
+        """
+        raise NotImplementedError
+
+    def create_message(self, recipient, content, related_objects, tag, template, **kwargs):
+        """
+        Create message which will be logged in the database.
+        :param recipient: email or phone number of the recipient
+        :param content: content of the message
+        :param related_objects: list of related objects that will be linked with the message using generic
+        relation
+        :param tag: string mark that will be saved with the message
+        :param template: template object from which content of the message was created
+        :param kwargs: extra attributes that will be saved with the message
+        """
+        message = self.model.objects.create(
+            recipient=recipient,
+            content=content,
+            tag=tag,
+            template=template,
+            template_slug=template.slug if template else None,
+            **kwargs
+        )
+        if related_objects:
+            message.related_objects.create_from_related_objects(*related_objects)
+        return message
+
+    def _get_backend_messages_map(self, messages):
+        backends_messages_map = defaultdict(list)
+        for message in messages:
+            backend = self.get_backend(recipient=message.recipient)
+            backends_messages_map[backend].append(message)
+        return backends_messages_map
+
+    def bulk_send_messages(self, messages):
+        """
+        Sends more messages together. If concrete backend provides send more messages at once the method
+        can be overridden
+        :param messages: list of messages
+        """
+        for backend, messages_for_backend in self._get_backend_messages_map(messages):
+            backend.publish_messages(messages_for_backend)
+
+    @transaction.atomic
```

Comment:
pozor na to atomic, když ti padne jedna zpráva tak se ti nesmí udělat už rollback na ty předchozí (ten atomic už tam byl?)